.. _WFoverview:

Here we need to describe input/output and how the subworkflows are integrated for each processing step

PREPROCESSING
=============

FETCH
#####
.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +------------------------+------------------------------------------------------------------------------------+------+---------------+---------------------------------------------+----------------+---------+
  | TOOL                   | DESCRIPTION                                                                        | ENV  | BIN           | LINK                                        | INPUT          | OUTPUT  |
  +========================+====================================================================================+======+===============+=============================================+================+=========+
  | SRAtools fasterq-dump  | The fasterq-dump tool extracts data in FASTQ- or FASTA-format from SRA-accessions  | sra  | fasterq-dump  | `sra <https://github.com/ncbi/sra-tools>`_  | SRA accession  | FASTQ   |
  +------------------------+------------------------------------------------------------------------------------+------+---------------+---------------------------------------------+----------------+---------+


BASECALL
########

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+---------------------+------------------------------------------------------------------------+--------+---------+
  | TOOL  | DESCRIPTION                                                                                                                                                                                                                                                                | ENV    | BIN                 | LINK                                                                   | INPUT  | OUTPUT  |
  +=======+============================================================================================================================================================================================================================================================================+========+=====================+========================================================================+========+=========+
  | Guppy | Data processing toolkit that contains Oxford Nanopore’s basecalling algorithms, and several bioinformatic post-processing features, such as barcoding/demultiplexing, adapter trimming, and alignment. Needs to be installed locally as no ``conda`` version is available  | guppy  | $PATH_TO_LOCAL_BIN  | `guppy <https://nanoporetech.com/nanopore-sequencing-data-analysis>`_  | FAST5  | FASTQ   |
  +-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+---------------------+------------------------------------------------------------------------+--------+---------+
  

QUALITY CONTROL I
##################

This workflow step can be run as preprocessing step if none of the processing workflows is defined in the config.json.

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +----------------------------+------------------------------------------------------------+---------+---------+-------------------------------------------------------------------------+------------+-----------+
  | TOOL                       | DESCRIPTION                                                | ENV     | BIN     | LINK                                                                    | INPUT      | OUTPUT    |
  +============================+============================================================+=========+=========+=========================================================================+============+===========+
  | FASTQC (includes MULTIQC)  | A quality control tool for high throughput sequence data.  | fastqc  | fastqc  | `fastqc <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/>`_  | FASTQ/BAM  | ZIP/HTML  |
  +----------------------------+------------------------------------------------------------+---------+---------+-------------------------------------------------------------------------+------------+-----------+
  

PROCESSING
==========

QUALITY CONTROL II
###################

If any of the below listed processing steps is defined in the config.json, quality control will run for all generated output files if enabled. 

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +----------------------------+------------------------------------------------------------+---------+---------+-------------------------------------------------------------------------+------------+-----------+
  | TOOL                       | DESCRIPTION                                                | ENV     | BIN     | LINK                                                                    | INPUT      | OUTPUT    |
  +============================+============================================================+=========+=========+=========================================================================+============+===========+
  | FASTQC (includes MULTIQC)  | A quality control tool for high throughput sequence data.  | fastqc  | fastqc  | `fastqc <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/>`_  | FASTQ/BAM  | ZIP/HTML  |
  +----------------------------+------------------------------------------------------------+---------+---------+-------------------------------------------------------------------------+------------+-----------+


Trimming
########

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+--------------+----------------------------------------------------------------------------------+--------+----------------+
  | TOOL         | DESCRIPTION                                                                                                                                                                                                              | ENV         | BIN          | LINK                                                                             | INPUT  | OUTPUT         |
  +==============+==========================================================================================================================================================================================================================+=============+==============+==================================================================================+========+================+
  | trim_galore  | A wrapper tool around Cutadapt and FastQC to consistently apply quality and adapter trimming to FastQ files, with some extra functionality for MspI-digested RRBS-type (Reduced Representation Bisufite-Seq) libraries.  | trimgalore  | trim_galore  | `trimgalore <https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/>`_  | FASTQ  | TRIMMED_FASTQ  |
  +--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+--------------+----------------------------------------------------------------------------------+--------+----------------+
  | cutadapt     | Cutadapt finds and removes adapter sequences, primers, poly-A tails and other types of unwanted sequence from your high-throughput sequencing reads.                                                                     | cutadapt    | cutadapt     | `cutadapt <https://cutadapt.readthedocs.io/en/stable/>`_                         | FASTQ  | TRIMMED_FASTQ  |
  +--------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+--------------+----------------------------------------------------------------------------------+--------+----------------+


Mapping
#######

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +--------------+----------------------------------------------------------------------------------------------------------------------------+----------------------+-------------+---------------------------------------------------------------------+----------------------+-------------+
  | TOOL         | DESCRIPTION                                                                                                                | ENV                  | BIN         | LINK                                                                | INPUT                | OUTPUT      |
  +==============+============================================================================================================================+======================+=============+=====================================================================+======================+=============+
  | HISAT2       | HISAT2 is a fast and sensitive alignment program                                                                           | hisat2               | hisat2      | `hisat2 <http://daehwankimlab.github.io/hisat2/manual/>`_           | FASTQ/TRIMMED_FASTQ  | SAM.gz/BAM  |
  +--------------+----------------------------------------------------------------------------------------------------------------------------+----------------------+-------------+---------------------------------------------------------------------+----------------------+-------------+
  | STAR         | Spliced Transcripts Alignment to a Reference                                                                               | star                 | star        | `star <https://github.com/alexdobin/STAR>`_                         | FASTQ/TRIMMED_FASTQ  | SAM.gz/BAM  |
  +--------------+----------------------------------------------------------------------------------------------------------------------------+----------------------+-------------+---------------------------------------------------------------------+----------------------+-------------+
  | Segemehl2|3  | Segemehl is a software to map short sequencer reads to reference genomes.                                                  | segemehl2/segemehl3  | segemehl.x  | `segemehl <https://www.bioinf.uni-leipzig.de/Software/segemehl/>`_  | FASTQ/TRIMMED_FASTQ  | SAM.gz/BAM  |
  +--------------+----------------------------------------------------------------------------------------------------------------------------+----------------------+-------------+---------------------------------------------------------------------+----------------------+-------------+
  | BWA          | BWA is a software package for mapping low-divergent sequences against a large reference genome                             | bwa                  | bwa mem     | `bwa <http://bio-bwa.sourceforge.net/>`_                            | FASTQ/TRIMMED_FASTQ  | SAM.gz/BAM  |
  +--------------+----------------------------------------------------------------------------------------------------------------------------+----------------------+-------------+---------------------------------------------------------------------+----------------------+-------------+
  | Minimap2     | Minimap2 is a versatile sequence alignment program that aligns DNA or mRNA sequences against a large reference database.   | minimap              | minimap2    | `minimap <https://github.com/lh3/minimap2>`_                        | FASTQ/TRIMMED_FASTQ  | SAM.gz/BAM  |
  +--------------+----------------------------------------------------------------------------------------------------------------------------+----------------------+-------------+---------------------------------------------------------------------+----------------------+-------------+


DEDUP
#####

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +---------------+----------------------------------------------------------------------------------------------------------------------------------------------------+-----------+------------+----------------------------------------------------------------------------------------------------+----------------------+------------+
  | TOOL          | DESCRIPTION                                                                                                                                        | ENV       | BIN        | LINK                                                                                               | INPUT                | OUTPUT     |
  +===============+====================================================================================================================================================+===========+============+====================================================================================================+======================+============+
  | UMI-tools     | UMI-tools contains tools for dealing with Unique Molecular Identifiers (UMIs)/Random Molecular Tags (RMTs) and single cell RNA-Seq cell barcodes.  | umitools  | umi_tools  | `umitools <https://umi-tools.readthedocs.io/en/latest/>`_                                          | FASTQ/TRIMMED_FASTQ  | FASTQ/BAM  |
  +---------------+----------------------------------------------------------------------------------------------------------------------------------------------------+-----------+------------+----------------------------------------------------------------------------------------------------+----------------------+------------+
  | Picard tools  | A better duplication marking algorithm that handles all cases including clipped and gapped alignments.                                             | picard    | picard     | `picard <https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard->`_  | BAM                  | BAM        |
  +---------------+----------------------------------------------------------------------------------------------------------------------------------------------------+-----------+------------+----------------------------------------------------------------------------------------------------+----------------------+------------+


POSTPROCESSING
==============

Read-Counting and Quantification
################################

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +----------------+-----------------------------------------------------------------------------------------------------------------------+-------------+----------------+------------------------------------------------------------------+----------------------+---------+
  | TOOL           | DESCRIPTION                                                                                                           | ENV         | BIN            | LINK                                                             | INPUT                | OUTPUT  |
  +================+=======================================================================================================================+=============+================+==================================================================+======================+=========+
  | FeatureCounts  | A software program developed for counting reads to genomic features such as genes, exons, promoters and genomic bins  | countreads  | featureCounts  | `featurecounts <http://subread.sourceforge.net/>`_               | BAM/FASTQ            | TEXT    |
  +----------------+-----------------------------------------------------------------------------------------------------------------------+-------------+----------------+------------------------------------------------------------------+----------------------+---------+
  | Salmon         | Salmon is a tool for wicked-fast transcript quantification from RNA-seq data.                                         | salmon      | salmon         | `salmon <https://salmon.readthedocs.io/en/latest/salmon.html>`_  | FASTQ/TRIMMED_FASTQ  | TEXT    |
  +----------------+-----------------------------------------------------------------------------------------------------------------------+-------------+----------------+------------------------------------------------------------------+----------------------+---------+


Differential Analyses
#####################

.. table:: 
  :class: tight-table

  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | Tool      | Analysis                            | Filtering        | Normalization   | Distribution   | Testing                         | Significance   | Results Table                                        | further                                 | SigTables                               | Clustering        | further                                                           | Rmd   |
  +===========+=====================================+==================+=================+================+=================================+================+======================================================+=========================================+=========================================+===================+===================================================================+=======+
  | edgeR     | Differential Gene Expression        | filterByExpr()   | TMM             | NB             | Fisher’s exact test             | pValue, LFC    | results, sorted-results                              | normalized                              | Sig, SigUP, SigDOWN                     | MDS-plot          | BCV, QLDisp, MD(per comparison)                                   | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | edgeR     | Differential Exon Usage             | filterByExpr()   | TMM             | NB             | Fisher’s exact test             | pValue, LFC    | results                                              | normalized                              |                                         | MDS-plot          | BCV, QLDisp, MD(per comparison)                                   | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | edgeR     | Differential Alternative Splicing   | filterByExpr()   | TMM             | NB             | Simes, gene-level, exon-level   | pValue, LFC    | results(diffSpliceExonTest, Simes-Test, Gene-Test)   |                                         | Sig, SigUP, SigDOWN                     | MDS-plot          | BCV, QLDisp, MD(per comparison), topSpliceSimes-plots(per Gene)   | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | DESeq2    | Differential Gene Expression        | RowSums >= 10    | RLE             | NB             | Wald test                       | pValue, LFC    | results                                              | rld, vsd, results(per comparison)       | Sig, SigUP, SigDOWN                     | PCA               | Heatmaps, MA(per comparison), VST-and-log2                        | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | DEXSeq    | Differential Exon Usage             | RowSums >= 10    | RLE             | Cox-Reid       | likelihood ratio test           |                |                                                      |                                         |                                         |                   |                                                                   |       |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | DEXSeq    | Differential Transcript Usage       | dmFilter()       | RLE             | Cox-Reid       | likelihood ratio test           | pValue         | results                                              |                                         |                                         |                   |                                                                   | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | DIEGO     | Differential Alternative Splicing   |                  |                 |                | Mann-Whitney U test             | pValue         | results                                              |                                         | Sig                                     | Dendrogram-plot   |                                                                   | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  | DRIMSeq   | Differential Transcript Usage       | dmFilter()       |                 | DM             |                                 | pValue, LFC    | results(transcript, genes)                           | Proportions-table, genewise precision   | Sig, SigUP, SigDOWN (transcipt, gene)   |                   | FeatPerGene, precision, Pvalues (per comparison)                  | ✓     |
  +-----------+-------------------------------------+------------------+-----------------+----------------+---------------------------------+----------------+------------------------------------------------------+-----------------------------------------+-----------------------------------------+-------------------+-------------------------------------------------------------------+-------+
  
TRACKS
###############

This workflow generates trackdb.txt files and bigwig files which can be used to create UCSC track hubs. However, bigwigs can of course be used for other genome viewers as well.

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +-------+----------------------------------------------------------------------------------------------------------+------+------+------------------------------------------------------------------------------+--------+--------------+
  | TOOL  | DESCRIPTION                                                                                              | ENV  | BIN  | LINK                                                                         | INPUT  | OUTPUT       |
  +=======+==========================================================================================================+======+======+==============================================================================+========+==============+
  | UCSC  | Track hubs are web-accessible directories of genomic data that can be viewed on the UCSC Genome Browser  | ucsc | ucsc | `ucsc <https://genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html#Intro>`_  | BAM    | BIGWIG/HUBS  |
  +-------+----------------------------------------------------------------------------------------------------------+------+------+------------------------------------------------------------------------------+--------+--------------+


PEAKS
#####

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +----------+----------------------------------------------------------------------------------------------------------------------+----------+----------+--------------------------------------------------------------------+--------+------------------+
  | TOOL     | DESCRIPTION                                                                                                          | ENV      | BIN      | LINK                                                               | INPUT  | OUTPUT           |
  +==========+======================================================================================================================+==========+==========+====================================================================+========+==================+
  | Piranha  | Piranha is a peak-caller for CLIP- and RIP-Seq data.                                                                 | piranha  | piranha  | `piranha <http://smithlabresearch.org/software/piranha/>`_         | BAM    | BED/BEDG/BIGWIG  |
  +----------+----------------------------------------------------------------------------------------------------------------------+----------+----------+--------------------------------------------------------------------+--------+------------------+
  | MACS     | Model-based Analysis of ChIP-Seq (MACS), for identifying transcript factor binding sites.                            | macs     | macs     | `macs <https://github.com/macs3-project/MACS>`_                    | BAM    | BED/BEDG/BIGWIG  |
  +----------+----------------------------------------------------------------------------------------------------------------------+----------+----------+--------------------------------------------------------------------+--------+------------------+
  | SciPhy   | Software for cyPhyRNA-Seq Data analysis                                                                              | scyphy   | piranha  | `cyphyRNA-Seq <https://doi.org/10.1080/15476286.2021.1999105>`_    | BAM    | BED/BEDG/BIGWIG  |
  +----------+----------------------------------------------------------------------------------------------------------------------+----------+----------+--------------------------------------------------------------------+--------+------------------+
  | Peaks    | Slinding window peak finding tool for quick assessment of peaks. UNPUBLISHED, recommended for initial scanning only  | peaks    | peaks    | `ttp <https://www.embopress.org/doi/full/10.15252/msb.20156628>`_  | BAM    | BED/BEDG/BIGWIG  |
  +----------+----------------------------------------------------------------------------------------------------------------------+----------+----------+--------------------------------------------------------------------+--------+------------------+


CIRCS
###############

CIRI2 needs to be installed locally.

.. table:: 
  :widths: 10, 40, 10, 10, 10, 10, 10
  :class: tight-table

  +-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+--------------------+-----------------------------------------------------------------------+--------+------------------+
  | TOOL  | DESCRIPTION                                                                                                                                                                                             | ENV    | BIN                | LINK                                                                  | INPUT  | OUTPUT           |
  +=======+=========================================================================================================================================================================================================+========+====================+=======================================================================+========+==================+
  | CIRI2 | CIRI (circRNA identifier) is a novel chiastic clipping signal based algorithm,which can unbiasedly and accurately detect circRNAs from transcriptome data by employing multiple filtration strategies.  | ciri2  | $Path_to_CIRI2.pl  | `ciri2 <https://ciri-cookbook.readthedocs.io/en/latest/CIRI2.html>`_  | BAM    | BED/BEDG/BIGWIG  |
  +-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+--------------------+-----------------------------------------------------------------------+--------+------------------+
